### PR TITLE
fix: Weave Object's Ops should be saved as ops!

### PR DESCRIPTION
**Pull request by tssweeney**
1. Fixes issue where bound ops are not actually bound to their parent self
2. Fixes object publishing to correctly publish the op

### CODE REVIEW

File: weave/weave_client.py [hunk 1]
```diff
@@ -739,8 +739,11 @@ def _save_op(self, op: Op, name: Optional[str] = None) -> Ref:
 
         if inspect.ismethod(op):
             # this "func" is both a method AND an Op, but we need to cast to get dot access to its attributes
-            op = cast(op, Op)  # type: ignore
-        op.ref = op_def_ref  # type: ignore
+            op = cast(Op, op)
```
Comment (tssweeney) [inline weave/weave_client.py:742]: this was flipped, which resulted in the ref not actually getting assigned to the object!

File: weave/trace/op.py [hunk 2]
```diff
@@ -350,7 +356,7 @@ def wrapper(*args: Any, **kwargs: Any) -> Any:
             wrapper.calls = partial(calls, wrapper)  # type: ignore
 
             wrapper.__call__ = wrapper  # type: ignore
-            wrapper.__self__ = wrapper  # type: ignore
+            wrapper.__self__ = None  # type: ignore
```
Comment (tssweeney) [inline weave/trace/op.py:359]: `__self__` was completely unused before - coopting it to mean "the self argument for ops that are gotten"

File: weave/trace/op.py [hunk 3]
```diff
@@ -372,6 +372,8 @@ def maybe_bind_method(func: Callable, self: Any = None) -> Union[Callable, Metho
     If self is None, return the function as is.
     """
     if (sig := inspect.signature(func)) and sig.parameters.get("self"):
+        if inspect.ismethod(func) and id(func.__self__) != id(self):
+            raise ValueError("Cannot re-bind a method to an new object")
```
Comment (andrewtruong) [inline weave/trace/op.py:376]: Should you be able to?

File: weave/weave_client.py (hunk 1 above)
Comment (andrewtruong) [reply weave/weave_client.py:742]: Wow, great catch!

**Review by andrewtruong — APPROVED** [review]

File: weave/weave_client.py (hunk 1 above)
Comment (andrewtruong) [inline weave/weave_client.py:742]: This is no longer required.  The cast was only done to assign to `ref`
```suggestion

```

File: weave/trace/op.py (hunk 3 above)
Comment (tssweeney) [reply weave/trace/op.py:376]: I can't think of a case where hitting this path is intentional, but seems more like a programming error. Maybe if we have a real use case for it we can remove

**Pull request merged**